tensorflow-utils/src/main/java/org/tensorflow/util/ndarray/index/Index.java

```diff
@@ -0,0 +1,77 @@
+/*
+ Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ =======================================================================
+ */
+package org.tensorflow.util.ndarray.index;
+
+import org.tensorflow.util.ndarray.NdArray;
+import org.tensorflow.util.ndarray.impl.dimension.Dimension;
+
+/**
+ * An index used for slicing a view out of an N-dimensional array.
+ *
+ * <p>A slice, i.e. a reduced view, of an N-dimensional array is obtain by calling
+ * {@link NdArray#slice(Index...)}, given a list of indices
+ * that select which elements on a given dimension should be included/excluded
+ * from that view.
+ */
+public interface Index {
```

Comment:
Is there a way to make this interface functional or more constant? I'm a little worried that the JVM won't be able to optimise all these small objects properly (until they become value types). Maybe we could make them `ValueBasedClass`es? They provide conceptually similar functionality to `VarHandle` from Java 9, which Paul wrote, so maybe his input would be useful on these classes.

Reply:
Ok, let's do this as a follow up.